hibernate-core/src/test/java/org/hibernate/test/tenantid/TenantIdTest.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.test.tenantid;
+
+import org.hibernate.boot.SessionFactoryBuilder;
+import org.hibernate.boot.spi.MetadataImplementor;
+import org.hibernate.context.spi.CurrentTenantIdentifierResolver;
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.testing.orm.junit.DomainModel;
+import org.hibernate.testing.orm.junit.ServiceRegistry;
+import org.hibernate.testing.orm.junit.SessionFactory;
+import org.hibernate.testing.orm.junit.SessionFactoryProducer;
+import org.hibernate.testing.orm.junit.SessionFactoryScope;
+import org.hibernate.testing.orm.junit.Setting;
+import org.junit.jupiter.api.Test;
+
+import static org.hibernate.cfg.AvailableSettings.HBM2DDL_DATABASE_ACTION;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+@SessionFactory
+@DomainModel(annotatedClasses = Account.class)
+@ServiceRegistry(
+        settings = {
+                @Setting(name = HBM2DDL_DATABASE_ACTION, value = "create-drop")
+        }
+)
+public class TenantIdTest implements SessionFactoryProducer {
+
+    String currentTenant;
+
+    @Override
+    public SessionFactoryImplementor produceSessionFactory(MetadataImplementor model) {
+        final SessionFactoryBuilder sessionFactoryBuilder = model.getSessionFactoryBuilder();
+        sessionFactoryBuilder.applyCurrentTenantIdentifierResolver( new CurrentTenantIdentifierResolver() {
```

Comment:
It would be awesome if we could update `org.hibernate.boot.registry.selector.internal.StrategySelectorImpl#resolveDefaultableStrategy(java.lang.Class<T>, java.lang.Object, java.util.concurrent.Callable<T>)` to allow instantiating a "bean" i.e. allow dependency injection to happen into the resolver instance.

Reply:
FTR, moved to https://hibernate.atlassian.net/browse/HHH-14822